diff --git a/src/dwarf/Gparser.c b/src/dwarf/Gparser.c
index 3a47255..59af68c 100644
--- a/src/dwarf/Gparser.c
+++ b/src/dwarf/Gparser.c
@@ -849,10 +849,16 @@ uncached_dwarf_find_save_locs (struct dwarf_cursor *c)
     }
 
   if ((ret = create_state_record_for (c, &sr, c->ip)) < 0)
-    return ret;
+    {
+      put_unwind_info (c, &c->pi);
+      return ret;
+    }
 
   if ((ret = apply_reg_state (c, &sr.rs_current)) < 0)
-    return ret;
+    {
+      put_unwind_info (c, &c->pi);
+      return ret;
+    }
 
   put_unwind_info (c, &c->pi);
   return 0;
